Add frontend mode -build-module-from-parseable-interface

jrose-apple opened the pull request:
Makes it easier to test the caching behavior, and may also be useful for "prebuilding" swiftinterfaces in the future, or having the Driver kick off a bunch of separate builds as proper tasks.

jrose-apple committed:
[ParseableInterface] Split up creation of the sub-CompilerInvocation

The goal here is to separate the parts that compute an output file
name from the parts that do the actual compilation, so that we can
test the swiftinterface -> swiftmodule behavior more directly. No
functionality change in this commit; the next will take advantage
of the refactoring.

jrose-apple commented:
Tagged @davidungar for the FrontendTool changes, since he's done plenty of stuff there, and @harlanhaskins and @graydon for the ParseableInterfaceModuleLoader changes.

@swift-ci Please test

davidungar commented on lib/FrontendTool/FrontendTool.cpp:
```diff
@@ -1749,9 +1749,22 @@ int swift::performFrontend(ArrayRef<const char *> Args,
+
+  if (Invocation.getFrontendOptions().RequestedAction ==
+      FrontendOptions::ActionType::BuildParseableInterface) {
+    const auto &InputsAndOutputs =
```
For the sake of symmetry, I'm thinking the code would read better by factoring out then then-block here into a function so that there is symmetry with the else-half. It's also odd-seeming on the surface that performCompile handles every other action, which is how it seems to be just from looking at this change.

jrose-apple commented:
Answer: that is in fact exactly what I did.

harlanhaskins commented:
Oh, didn't see that the flag is just `to-module`. What about `-build-module-from-parseable-interface`?

jrose-apple committed:
Add frontend mode -build-module-from-parseable-interface

Makes it easier to test the caching behavior, and may also be useful
for "prebuilding" swiftinterfaces in the future, or having the Driver
kick off a bunch of separate builds as proper tasks.

jrose-apple commented:
Updated per review feedback!

@swift-ci Please test

harlanhaskins reviewed (approved):
Thanks!

davidungar reviewed (commented):
Sorry I didn't get deeper into this; looks good to me!
